cardinal/cardinal.go

```diff
@@ -0,0 +1,346 @@
+		if !eris.Is(err, redis.ErrNoSchemaFound) {
+			return err
+		}
+	} else {
+		valid, err := types.IsComponentValid(t, storedSchema)
+		if err != nil {
+			return err
+		}
+		if !valid {
+			return eris.Errorf("Component: %s does not match the type stored in the db", c.Name())
+		}
+	}
+
+	err = w.redisStorage.SetSchema(c.Name(), c.GetSchema())
+	if err != nil {
+		return err
+	}
+	w.nameToComponent[t.Name()] = c
+	w.isComponentsRegistered = true
+	return nil
+}
+
+func MustRegisterComponent[T types.Component](w *World) {
+	err := RegisterComponent[T](w)
+	if err != nil {
+		panic(err)
+	}
+}
+
+// RegisterMessages adds the given messages to the game world. HTTP endpoints to queue up/execute these
+// messages will automatically be created when StartGame is called. This Register method must only be called once.
+func RegisterMessages(w *World, msgs ...types.Message) error {
+	if w.WorldState != WorldStateInit {
+		return eris.Errorf(
+			"engine state is %s, expected %s to register messages",
+			w.WorldState,
+			WorldStateInit,
+		)
+	}
+	return w.msgManager.RegisterMessages(msgs...)
+}
+
+func RegisterQuery[Request any, Reply any](
+	world *World,
+	name string,
+	handler func(wCtx engine.Context, req *Request) (*Reply, error),
+	opts ...QueryOption[Request, Reply],
+) error {
+	if world.WorldState != WorldStateInit {
+		panic("cannot register queries after loading game state")
+	}
+
+	if _, ok := world.nameToQuery[name]; ok {
+		return eris.Errorf("query with name %s is already registered", name)
+	}
+
+	q, err := NewQueryType[Request, Reply](name, handler, opts...)
+	if err != nil {
+		return err
+	}
+
+	world.registeredQueries = append(world.registeredQueries, q)
+	world.nameToQuery[q.Name()] = q
+
+	return nil
+}
+
+// Create creates a single entity in the world, and returns the id of the newly created entity.
+// At least 1 component must be provided.
+func Create(wCtx engine.Context, components ...types.Component) (types.EntityID, error) {
+	// Error if the context is read only
+	if wCtx.IsReadOnly() {
+		return 0, ErrEntityMutationOnReadOnly
+	}
+
+	id, err := CreateMany(wCtx, 1, components...)
+	if err != nil {
+		return 0, logAndPanic(wCtx, err)
+	}
+
+	return id[0], nil
+}
+
+// CreateMany creates multiple entities in the world, and returns the slice of ids for the newly created
+// entities. At least 1 component must be provided.
+func CreateMany(wCtx engine.Context, num int, components ...types.Component) ([]types.EntityID, error) {
+	if !wCtx.IsWorldReady() {
+		return nil, eris.Wrap(ErrEntitiesCreatedBeforeStartGame, "")
+	}
+
+	// Error if the context is read only
+	if wCtx.IsReadOnly() {
+		return nil, ErrEntityMutationOnReadOnly
+	}
+
+	// Get all component metadata for the given components
+	acc := make([]types.ComponentMetadata, 0, len(components))
+	for _, comp := range components {
+		c, err := wCtx.GetComponentByName(comp.Name())
+		if err != nil {
+			return nil, logAndPanic(wCtx, ErrComponentNotRegistered)
```

Comment:
This refactor will make keeping track of these non-deterministic errors harder. Any modifications to these state-accessing methods need to make sure logAndPanic is used.

Reply:
refactor-ed error handling to user defer handler to make this more  straight forward